docs/using/recipe_code/oracledb.dockerfile

```diff
@@ -0,0 +1,63 @@
+    apt-get clean && rm -rf /var/lib/apt/lists/*
+
+## Then install Oracle SQL Instant client, SQL+Plus, tools and JDBC.
+## Note: You may need to change the URL to a newer version.
+## See: https://www.oracle.com/es/database/technologies/instant-client/linux-x86-64-downloads.html
+RUN mkdir "/opt/oracle"
+WORKDIR "/opt/oracle"
+RUN wget --progress=dot:giga https://download.oracle.com/otn_software/linux/instantclient/2111000/oracle-instantclient-basiclite-21.11.0.0.0-1.el8.x86_64.rpm && \
+    alien -d --scripts oracle-instantclient-basiclite-21.11.0.0.0-1.el8.x86_64.rpm && \
+    dpkg -i oracle-instantclient-basiclite_21.11.0.0.0-2_amd64.deb && \
+    wget --progress=dot:giga https://download.oracle.com/otn_software/linux/instantclient/2111000/oracle-instantclient-sqlplus-21.11.0.0.0-1.el8.x86_64.rpm && \
+    alien -d  --scripts oracle-instantclient-sqlplus-21.11.0.0.0-1.el8.x86_64.rpm && \
+    dpkg -i oracle-instantclient-sqlplus_21.11.0.0.0-2_amd64.deb && \
+    wget --progress=dot:giga https://download.oracle.com/otn_software/linux/instantclient/2111000/oracle-instantclient-tools-21.11.0.0.0-1.el8.x86_64.rpm && \
+    alien -d --scripts oracle-instantclient-tools-21.11.0.0.0-1.el8.x86_64.rpm && \
+    dpkg -i oracle-instantclient-tools_21.11.0.0.0-2_amd64.deb && \
+    wget --progress=dot:giga https://download.oracle.com/otn_software/linux/instantclient/2111000/oracle-instantclient-jdbc-21.11.0.0.0-1.el8.x86_64.rpm && \
+    alien -d --scripts oracle-instantclient-jdbc-21.11.0.0.0-1.el8.x86_64.rpm && \
+    dpkg -i oracle-instantclient-jdbc_21.11.0.0.0-2_amd64.deb && \
+    # Remove temporary files to avoid issues with `fix-permissions` commands
+    rm "*.deb" && rm "*.rpm" && chown -R "${NB_UID}":"${NB_GID}" "${HOME}/.rpmdb"
+
+## Configure environment
+## Note: You may need to change the ORACLE_HOME path to a different version `.../oracle/21/...`.
+ENV ORACLE_HOME=/usr/lib/oracle/21/client64
+ENV PATH="${ORACLE_HOME}/bin:${PATH}"
+ENV LD_LIBRARY_PATH=${ORACLE_HOME}/lib
+ENV PATH="${LD_LIBRARY_PATH}:${PATH}"
+
+## (Optional) Add credentials for the Oracle Database server; files must be present on your root folder.
+WORKDIR /usr/lib/oracle/21/client64/lib/network/admin
+## Adding a wildcard `[]` on the last letter of the filename to avoid throwing an error if the file does not exist.
+## See: https://stackoverflow.com/questions/31528384/conditional-copy-add-in-dockerfile
+COPY cwallet.ss[o] ./
+COPY sqlnet.or[a] ./
+COPY tnsnames.or[a] ./
+
+## Switch back to jovyan.
+USER "${NB_UID}"
+
+## Change workdir back to jovyan.
+WORKDIR "${NB_UID}"
+
+## Install `oracledb` Python library to use Oracle SQL Instant Client with `--upgrade --user` options enabled
+
+# Alternatively, if you are using `conda` to add the custom Python environment and Jupyter kernel:
+# RUN "${CONDA_DIR}/envs/${CONDA_ENV}/bin/pip" install --no-cache-dir oracledb --upgrade --user && \
+#     fix-permissions "${CONDA_DIR}" && \
+#     fix-permissions "${HOME}"
+
```

Comment:
I suggest to remove this part. If you're using custom python environment, you can install this package during environment creation.

Reply:
Uhm... 🤔 

Oracle suggests to use these options '--upgrade --user' and I am not quite sure how you can specify that from a `environment.yml` file or similar.

Even in the original Docker file this recipe the `oracledb` library install is done on a separate command rather than during the environment creation.

Oracle drivers can be a bit tricky and I have opted to follow their suggestions.

> `python -m pip install oracledb --upgrade`
> If a binary package is not available for your platform, the source package will be downloaded instead. This will be compiled and the resulting binary installed.
> The --user option may be useful if you do not have permission to write to system directories
> ... 
Reference: https://python-oracledb.readthedocs.io/en/latest/user_guide/installation.html

In any case, I am bit confused, @mathbunnyru.

Reading the contribution guidelines and other recipes in the very same page, it does not seem like there is a clear guidelines about what is expected of the recipe code.

Apart from being valid and linted code, I made the assumption that the source code submitted is meant to be informative in the context of a Technical documentation page. Rather than a code that needs to be executed on a working project.

Reply:
In any case, feel free to add any more improvements to this recipe.

Installing and configuring Oracle Database drivers is not quite obvious from the information one can find on the internet using a search engine.

So I hope this contribution may help others to save time in doing so.

Many thanks for your comments and insights.

Reply:
Oracle suggests to use these options '--upgrade --user' and I am not quite sure how you can specify that from a environment.yml file or similar.

You don't.
But we suggest using mamba (where `--user` doesn't make sense).
And we're creating a new environment (where `--upgrade` will have no effect, because a package is not yet installed).

That's why I suggest removing commented code.